### ISSUE TITLE
Possible useCollectionData() and useDocumentData()

### ISSUE THREAD

**Issue by nmaves** [CSFrequency/react-firebase-hooks#15]
I have found the collectionData() from rxfire to be super useful.

Currently, your hooks return a snapshot.  While I don't find that very useful others might.  So what if we added a new useCollectionData(query, idField='id') to the api.

Here is an example of how I wrap the useCollection from this library.
```
export const useCollectionData = (query, idField = 'id') => {
  const { error, loading, value } = useCollection(query)

  const docs = []
  if (!loading) {
    value.forEach(snap => {
      docs.push({
        ...snap.data(),
        [idField]: snap.id,
      })
    })
  }

  return [docs, loading, error]
}
```

There are two things to notice.  The first is that I return a collection of documents **NOT** a snapshot.  The second is that I return an array just like useState() from React.  I find this very helpful when using multiple useCollections() with a Component.

The same feature could be added to the useDocument() hooks as well.

**Comment by chrisbianca** [comment]
@nmaves Thanks for the suggestion.  This makes a lot of sense, particularly given that we have something very similar already for the Realtime Database with `useObjectVal`, `useListKeys` and `useListVals`.

I've almost finished `useCollectionData` and `useDocumentData`, just need to do a bit of testing, but will update here when they're ready to release.

On the subject of the return of an array vs an object, there's a bit of conversation about this on https://github.com/CSFrequency/react-firebase-hooks/issues/5, but no conclusion yet about which is the best way forward.  I'd welcome your input on that topic.

**Comment by chrisbianca** [comment]
`useCollectionData` and `useDocumentData` have been added as part of the v1.2.0 release: https://github.com/CSFrequency/react-firebase-hooks/releases/tag/v1.2.0

Discussions on the array return can continue on #5 .

**Comment by nmaves** [comment]
Hey Chris,

Sorry I have not been around to give this a try.  

One thing I would have changed was the order of these params.  

```
export default <T>(
  query?: firestore.Query | null,
  options?: firestore.SnapshotListenOptions,
  idField?: string
)
```

I would have put the idField second as it is way more common to use with this function than the ListenOptions.

**Comment by nmaves** [comment]
Also, I don't think it would be a bad thing to return an empty [] as the default state for useCollectionData.

**Comment by nmaves** [comment]
@chrisbianca the more I think about it and use the hooks I really think we should use a default value for the idField.  It is so common to need the document id I would just default it for everyone to use.  I do like the idea from the other issue to combine the options into one object.  Thoughts?

**Comment by chrisbianca** [comment]
@nmaves Thanks for your comments.

As per my comments on #5, I do think combining both optional parameters into a single optional object is a cleaner approach that gets around this problem, and will get this changed in v2.

With regards to the default value for document ID - I'd like to follow what rxFire have done which is to leave it as an optional field with no default.  Using a default of say `id` would cause issues for those people that already add an `id` field to their document data, so you'd be forcing them to have to specify a fieldName that didn't clash even though they wouldn't be using it.

I'm unsure on the empty [] as well - are you suggesting this be the default state when no reference is supplied?  My concern is that it's then hard to differentiate between an empty data set, or no data set at all.

**Comment by nmaves** [comment]
It is not a huge deal in my application as I have already wrapped these hooks with my own to always provide the array syntax (🎉 coming in V2).  So this is also where I provide the default fieldName value of `id`.

As for the `[]` being the default state for a `useCollectionData()` I prefer this to undefined as we no longer have access to the snapshot with these data methods.

Normally you would do something like this to check for empty results.

```
var query = citiesRef.where('capital', '==', true).get()
  .then(snapshot => {
    if (snapshot.empty) {
      console.log('No matching documents.');
      return;
    }

    snapshot.forEach(doc => {
      console.log(doc.id, '=>', doc.data());
    });
  })
  .catch(err => {
    console.log('Error getting documents', err);
  });
```

Now that we are hiding that if there are no results I would expect an empty array.

So there are 3 states, **loading**, **empty**, and **results**.

The more I think about this I realized that I wanted this in places that I chose not to check the loading state.  I have code like `const [people] = useCollectionData(query)` and I call `.map` on that collection without checks.  Lazy I know.  I think I am now on the fence about this.  I could go either way.
